libcore/src/SimulationHelper.cpp

```diff
@@ -0,0 +1,218 @@
+    }
+
+    return {pedsChangedRoom, pedsNotRelocated};
+}
+
+std::vector<Pedestrian *> SimulationHelper::FindOutsidePedestrians(
+    const Building & building,
+    std::vector<Pedestrian *> & peds)
+{
+    std::vector<Pedestrian *> pedsOutside;
+    std::vector<Pedestrian *> newPeds;
+
+    std::partition_copy(
+        std::make_move_iterator(std::begin(peds)),
+        std::make_move_iterator(std::end(peds)),
+        std::back_inserter(pedsOutside),
+        std::back_inserter(newPeds),
+        [&building](auto const ped) -> bool {
+            auto transPassed = SimulationHelper::FindPassedDoor(building, *ped);
+            if(!transPassed.has_value()) {
+                return false;
+            }
+
+            //TODO maxDistance should depend on vmax
+            double maxDistance = 0.5;
+            return transPassed.value()->IsExit() && transPassed.value()->IsOpen() &&
+                   transPassed.value()->DistTo(ped->GetPos()) < maxDistance;
+        });
+    peds = std::move(newPeds);
+    return pedsOutside;
+}
+
+void SimulationHelper::UpdateFlowAtDoors(
+    Building & building,
+    const std::vector<Pedestrian *> & peds)
+{
+    for(const auto ped : peds) {
+        auto closestTransition = SimulationHelper::FindPassedDoor(building, *ped);
+
+        if(!closestTransition.has_value()) {
+            LOG_WARNING("Ped {} did not cross any transition", ped->GetID());
+            return;
+        }
+
+        //TODO check if exitindex is set and transition exists?
+        if(closestTransition.value()->GetUniqueID() !=
+           building.GetTransitionByUID(ped->GetExitIndex())->GetUniqueID()) {
+            LOG_WARNING(
+                "Ped {}: used an indented door {}, but wanted to go to {}.",
+                ped->GetID(),
+                closestTransition.value()->GetUniqueID(),
+                ped->GetExitIndex());
+            return;
+        }
+
+        closestTransition.value()->IncreaseDoorUsage(1, Pedestrian::GetGlobalTime());
+        closestTransition.value()->IncreasePartialDoorUsage(1);
+    }
+}
+
+std::optional<Transition *>
+SimulationHelper::FindPassedDoor(const Building & building, const Pedestrian & ped)
+{
+    std::vector<Transition *> transitions;
+
+    // TODO check if room exists?
+    auto room = (ped.GetRoomID() != -1) ? (building.GetAllRooms().at(ped.GetRoomID())) :
+                                          (building.GetAllRooms().at(ped.GetOldRoomID()));
+
+    for(auto const & [_, subroom] : room->GetAllSubRooms()) {
+        transitions.insert(
+            std::end(transitions),
+            std::begin(subroom->GetAllTransitions()),
+            std::end(subroom->GetAllTransitions()));
+    }
+
+    Line step{ped.GetLastPosition(), ped.GetPos()};
+    // TODO check for closed doors and distance?
+    auto passedTrans = std::find_if(
+        std::begin(transitions), std::end(transitions), [&step](auto const & trans) -> bool {
+            return trans->IntersectionWith(step) == 1;
```

Comment:
What is in the case that the lines somehow overlap?

Reply:
The function now also deals with the cases:
1. Current step ends on line: No passed door
2. Current step is on line: No passed door
3. Current step starts on line: Door passed
Added the corresponding tests